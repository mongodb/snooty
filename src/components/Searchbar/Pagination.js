import React, { useCallback, useMemo } from 'react';
import styled from '@emotion/styled';
import Button from '@leafygreen-ui/button';
import Icon from '@leafygreen-ui/icon';
import { uiColors } from '@leafygreen-ui/palette';
import { theme } from '../../theme/docsTheme';

const BUTTON_HEIGHT = theme.size.medium;
const BUTTON_WIDTH = '14px';
const ENABLED_COLOR = uiColors.gray.dark2;
const DISABLED_COLOR = uiColors.gray.light1;

const PaginationButton = styled(Button)`
  background-color: #fff;
  height: ${BUTTON_HEIGHT};
  padding: 0;
  width: ${BUTTON_WIDTH};
  z-index: 1;
  /* Below removes default hover effects from button */
  background-image: none;
  border: none;
  box-shadow: none;
  :before {
    display: none;
  }
  :after {
    display: none;
  }
`;

const PaginationButtonIcon = styled(Icon)`
  height: ${BUTTON_HEIGHT};
  left: 0;
  position: absolute;
  top: 0;
  width: ${BUTTON_WIDTH};
`;

const PaginationContainer = styled('div')`
  align-items: center;
  display: flex;
`;

const PaginationText = styled('p')`
  font-size: ${theme.size.default};
  margin: 0 ${theme.size.tiny};
`;

const Pagination = ({ currentPage, setCurrentPage, totalPages }) => {
  const decrementPage = useCallback(() => setCurrentPage(currentPage - 1), [currentPage, setCurrentPage]);
  const incrementPage = useCallback(() => setCurrentPage(currentPage + 1), [currentPage, setCurrentPage]);
  const canDecrementPage = useMemo(() => currentPage !== 1, [currentPage]);
  const canIncrementPage = useMemo(() => currentPage < totalPages, [currentPage, totalPages]);
  return (
    <PaginationContainer>
      <PaginationButton
        aria-label="Back Page"
        disabled={!canDecrementPage}
        glyph={<PaginationButtonIcon glyph="ChevronLeft" fill={canDecrementPage ? ENABLED_COLOR : DISABLED_COLOR} />}
        onClick={decrementPage}
        title="Back Page"
      />
      <PaginationText>
        <strong>
          {currentPage}/{totalPages}
        </strong>
      </PaginationText>
      <PaginationButton
        aria-label="Forward Page"
        disabled={!canIncrementPage}
        glyph={<PaginationButtonIcon glyph="ChevronRight" fill={canIncrementPage ? ENABLED_COLOR : DISABLED_COLOR} />}
        onClick={incrementPage}
        title="Forward Page"
      />
    </PaginationContainer>
  );
};

export default Pagination;
