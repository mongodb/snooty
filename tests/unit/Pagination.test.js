import React, { useState } from 'react';
import { shallow, mount } from 'enzyme';
import Pagination from '../../src/components/Searchbar/Pagination';

// Simple wrapper to add state control around the Pagination component
const PaginationController = ({ startPage, totalPages }) => {
  const [page, setPage] = useState(startPage);
  return <Pagination currentPage={page} totalPages={totalPages} setCurrentPage={setPage} />;
};

describe('Pagination', () => {
  it('renders pagination correctly', () => {
    const wrapper = shallow(<Pagination currentPage={1} totalPages={10} setCurrentPage={jest.fn} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not allow page 1 to be decremented', () => {
    const wrapper = shallow(<Pagination currentPage={1} totalPages={10} setCurrentPage={jest.fn} />);
    let paginationText = wrapper.find('PaginationText').text();
    expect(paginationText).toBe('1/10');
    const decrementButton = wrapper.find('PaginationButton').at(0);
    expect(decrementButton.props().disabled).toBeTruthy();
    decrementButton.simulate('click');
    paginationText = wrapper.find('PaginationText').text();
    // Should not allow for 1 to be decremented
    expect(paginationText).toBe('1/10');
  });

  it('should allow pages > 1 to be decremented', () => {
    const paginationWithController = mount(<PaginationController startPage={2} totalPages={5} />);
    let paginationText = paginationWithController.find('PaginationText').text();
    expect(paginationText).toBe('2/5');
    const decrementButton = paginationWithController.find('PaginationButton').at(0);
    expect(decrementButton.props().disabled).toBeFalsy();
    decrementButton.simulate('click');
    paginationText = paginationWithController.find('PaginationText').text();
    expect(paginationText).toBe('1/5');
  });

  it('should allow pages < total pages to be incremented', () => {
    const paginationWithController = mount(<PaginationController startPage={2} totalPages={5} />);
    let paginationText = paginationWithController.find('PaginationText').text();
    expect(paginationText).toBe('2/5');
    const incrementButton = paginationWithController.find('PaginationButton').at(1);
    expect(incrementButton.props().disabled).toBeFalsy();
    incrementButton.simulate('click');
    paginationText = paginationWithController.find('PaginationText').text();
    expect(paginationText).toBe('3/5');
  });

  it('should not allow the last page to be incremented', () => {
    const paginationWithController = mount(<PaginationController startPage={5} totalPages={5} />);
    let paginationText = paginationWithController.find('PaginationText').text();
    expect(paginationText).toBe('5/5');
    const incrementButton = paginationWithController.find('PaginationButton').at(1);
    expect(incrementButton.props().disabled).toBeTruthy();
    incrementButton.simulate('click');
    paginationText = paginationWithController.find('PaginationText').text();
    expect(paginationText).toBe('5/5');
  });
});
