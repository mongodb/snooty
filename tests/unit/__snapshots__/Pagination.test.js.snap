// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders pagination correctly 1`] = `
<PaginationContainer>
  <PaginationButton
    aria-label="Back Page"
    disabled={true}
    glyph={
      <ForwardRef(render)
        fill="#B8C4C2"
        glyph="ChevronLeft"
      />
    }
    onClick={[Function]}
    title="Back Page"
  />
  <PaginationText>
    <strong>
      1
      /
      10
    </strong>
  </PaginationText>
  <PaginationButton
    aria-label="Forward Page"
    disabled={false}
    glyph={
      <ForwardRef(render)
        fill="#3D4F58"
        glyph="ChevronRight"
      />
    }
    onClick={[Function]}
    title="Forward Page"
  />
</PaginationContainer>
`;
